.github/workflows/api.yml

```diff
@@ -0,0 +1,106 @@
+name: api
+
+on:
+  push:
+    paths:
+      - "apigo/**"
+
+jobs:
+  lint:
+    runs-on: ubuntu-18.04
+
+    steps:
+      - run: |
+          echo "noop"
+
+  build:
+    runs-on: ubuntu-18.04
+
+    outputs:
+      tags: ${{ steps.metadata.outputs.tags }}
+
+    steps:
+      - id: metadata
+        run: |
+          IMAGE=arkhn/river-api-go
+          # Always tag with first 8 of sha
+          TAGS="${IMAGE}:${GITHUB_SHA::8}"
+
+          # Has this workflow been triggered by a tag ?
+          if [[ "${GITHUB_REF}" == refs/tags/* ]]; then
+            VERSION="${GITHUB_REF#refs/tags/}"
+
+            # Is the tag on master and does it follow numeric versioning ?
+            if [[ "${{ github.event.base_ref }}" == refs/heads/master ]] && [[ "${VERSION}" =~ ^v[0-9]{1,3}(\.[0-9]{1,3}(\.[0-9]{1,3})?)?$ ]]; then
+              TAGS="${TAGS},${IMAGE}:latest"
+              VERSION="${VERSION#v}"
+            fi
+
+          # Has this workflow been triggered by a branch ?
+          elif [[ "${GITHUB_REF}" == refs/heads/* ]]; then
+            VERSION=$(echo ${GITHUB_REF#refs/heads/} | sed -r 's#/+#-#g')
+          fi
+
+          if [[ ! -z "${VERSION}" ]]; then
+            TAGS="${TAGS},${IMAGE}:${VERSION}"
+          fi
+
+          echo ::set-output name=version::${VERSION}
+          echo ::set-output name=tags::${TAGS}
+          echo ::set-output name=created::$(date -u +'%Y-%m-%dT%H:%M:%SZ')
+      - uses: actions/checkout@v2
+      - uses: docker/setup-qemu-action@v1
+      - uses: docker/setup-buildx-action@v1
+        with:
+          driver-opts: network=host
+      - uses: docker/build-push-action@v2
+        with:
+          file: ./apigo/Dockerfile
+          context: ./apigo
+          build-args: |
+            VERSION_NAME=${{ steps.metadata.outputs.version }}
+            VERSION_SHA=${GITHUB_SHA::8}
+          platforms: linux/amd64
+          outputs: type=docker,dest=image
+          tags: ${{ steps.metadata.outputs.tags }}
+          labels: |
+            com.arkhn.image.title=api
+            com.arkhn.image.description=${{ github.event.repository.description }}
+            com.arkhn.image.url=${{ github.event.repository.html_url }}
+            com.arkhn.image.source=${{ github.event.repository.clone_url }}
+            com.arkhn.image.version=${{ steps.metadata.outputs.version }}
+            com.arkhn.image.created=${{ steps.metadata.outputs.created }}
+            com.arkhn.image.revision=${{ github.sha }}
+            com.arkhn.image.licenses=${{ github.event.repository.license.spdx_id }}
+      - uses: actions/upload-artifact@v2
```

Comment:
Do you think we could use https://github.com/marketplace/actions/build-and-push-docker-images#leverage-github-cache instead of artifacts?

Reply:
sounds cool

Reply:
https://docs.github.com/en/free-pro-team@latest/actions/guides/storing-workflow-data-as-artifacts#passing-data-between-jobs-in-a-workflow :

> Use caching when you want to reuse files that don't change often between jobs or workflow runs.

In our case, images builded with different workflows are never the same. Though sharing the builded image will save us time, compared to the naive approach of building the image whenever we need it, we are not primarely interested in caching.

Artifacts are the result of a computation we want to manipulate and have access to. They are not meant to speed up things. For instance we can retrieve a builded image and debug it manually.
Caches are used to speed up operations by saving predictable outputs. They are not meant to be access in other way. They can be removed, cleared or absent, while artifacts cannot.
You can have one or both, but for different usecases, different implementations. Caches will have less available size and shorter retention policy among other things.
These concepts are not restricted to github actions, you'll find the same thing in GitLabCI or more generally in CS.

So no for "can we use instead cache instead of artifacts?". But we could add cache to optimize (though building once is enough for now).

Reply:
Fine by me 👍